src/main/ApplicationImpl.h

```diff
@@ -116,6 +116,8 @@ class ApplicationImpl : public Application
 
     virtual AbstractLedgerTxnParent& getLedgerTxnRoot() override;
 
+    virtual void actBeforeDBSchemaUpgrade() override;
```

Comment:
Do you think this should be a virtual function of `Database` instead of `Application`? Seems like separation of concerns.

Reply:
The idea was that Application would be the one class which dictated where its clients could inject code.  If ```actBeforeDBSchemaUpgrade()``` were to move to the Database class, how are you envisioning clients creating Application instances that injected code into Database methods?

Reply:
I would add a virtual function `ApplicationImpl::createDatabase` (see `createInvariantManager`, `createOverlayManager`, and `createOverlayManager` for reference) and overload this in the derived class. Then you could use basically the same injection mechanism as you have here, where the derived `Application` takes a functor which passes it to the derived `Database` upon construction, and the functor is called before schema upgrade.

I admit it's a bit less direct. It has the disadvantage of requiring a bit more boiler-plate. But it has the advantage of having better separation of concerns and being more composable. I'm not sure how much it matters here, but I figured we should at least discuss it.

Reply:
Aha -- I had briefly mulled over trying to work out how to do that, but I was afraid it would come across as too heavyweight, and didn't get as far as noticing that we had precedent!

Though the result is more verbose, "having better separation of concerns and being more composable" is music to my ears, so I've taken your suggestion.  Does the result look sane?